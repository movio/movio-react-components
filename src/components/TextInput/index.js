// @flow
import React, { Component } from 'react';
import classnames from 'classnames';

import styles from './textInput.css';

type Props = {
  inputType: string,
  name: string,
  onChange: (value: string, event: SyntheticInputEvent) => void,
  placeholder: string,
  className: string,
  disabled: boolean,
  label?: string,
};

type State = {
  value: string,
};

class TextInput extends Component {
  props: Props;
  state: State;

  constructor(props: Props, context: *) {
    super(props, context);
    this.state = {
      value: '',
    };
  }

  static defaultProps = {
    inputType: 'text',
    placeholder: '...',
    disabled: false,
    className: '',
  };

  handleChange = (event: SyntheticInputEvent) => {
    const { onChange } = this.props;
    const value = event.target.value;
    onChange(value, event);
    this.setState({
      value: event.target.value,
    });
  };

  renderLabel = (label: string, name: string) => (
    <label className={styles.label} htmlFor={name}>{label}</label>
  );

  render() {
    const {
      label,
      inputType,
      placeholder,
      className,
      disabled,
      name,
    } = this.props;
    const { value } = this.state;
    const classList = classnames({
      [className]: className && !disabled,
      [styles.container]: true,
    });

    return (
      <fieldset className={classList}>
        {label && this.renderLabel(label, name)}
        <input
          id={name}
          className={styles.input}
          type={inputType}
          placeholder={placeholder}
          onChange={e => this.handleChange(e)}
          value={value}
          disabled={disabled}
        />
      </fieldset>
    );
  }
}

export default TextInput;
export { styles };
