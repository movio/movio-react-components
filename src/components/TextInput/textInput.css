@import "variables.css";

:root {
  --border-radius: 3px;
  --color-bg: var(--color-neutral-f);
  --normal-shadow: inset 0 0 0 2px transparent;
  --focus-shadow: inset 0 0 0 2px var(--color-primary-8);
  --color-bg-disabled: var(--color-neutral-d);
  --color-text-disabled: var(--color-neutral-b);
  --font-family: var(--font-primary);

  --disabled-styles: {
    cursor: not-allowed;
    background-color: var(--color-bg-disabled);
    color: var(--color-text-disabled);
  }
}

@custom-selector :--enter :focus:enabled, :active:enabled;

.container {
  border: none;
  outline: none;
  margin: 0;
  padding: 0;
  font-family: var(--font-family);
}

.input {
  font-family: var(--font-family);
  border: none;
  outline: none;
  font-size: 1rem;
  line-height: 2.5rem;
  height: 2.5rem;
  border-radius: var(--border-radius);
  padding: 0 1rem;
  background: var(--color-bg);
  transition: all 0.3s;
  box-shadow: var(--normal-shadow);
  width: 100%;

  @nest &:--enter {
    background: white;
    box-shadow: var(--focus-shadow);
  }

  @nest &:disabled {
    @apply --disabled-styles;
  }
}

.disabled {
  @apply --disabled-styles;
}

.label {
  margin: 0 0 0.25rem;
  display: block;
  font-size: 1rem;
}
