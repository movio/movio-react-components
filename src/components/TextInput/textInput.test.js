import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import TextInput, { styles } from './index';

const noop = () => {};

describe('<TextInput />', () => {
  it('should render one <input> element', () => {
    const component = <TextInput name="sample-input" onChange={noop} />;
    const wrapper = shallow(component);
    expect(wrapper.find('input')).toHaveLength(1);
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot('text-input-basic');
  });

  it('should optionally render a <label> element, if provided', () => {
    const component = <TextInput name="sample-input" onChange={noop} label="Label" />;
    const wrapper = shallow(component);
    expect(wrapper.find('label')).toHaveLength(1);
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot('text-input-label');
  });

  it('should append a className to the default, if provided', () => {
    const component = <TextInput name="sample-input" className="test-class" onChange={noop} />;
    const wrapper = shallow(component);
    expect(wrapper.find('input').prop('className')).toEqual(`test-class ${styles.input}`);
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot('text-input-custom-classname');
  });

  it('should use the default className from the stylesheet, if not provided', () => {
    const component = <TextInput name="sample-input" onChange={noop} />;
    const wrapper = shallow(component);
    expect(wrapper.find('input').prop('className')).toEqual(styles.input);
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot('text-input-default-classname');
  });

  it('should execute input prop onChange when the input value changes', () => {
    const spy = jest.fn(v => v);
    const synthEv = { target: { value: 'a' } };
    const component = <TextInput name="sample-input" onChange={spy} />;
    const wrapper = shallow(component);
    expect(spy.mock.calls).toHaveLength(0);
    wrapper.find('input').simulate('change', synthEv);
    expect(spy.mock.calls).toHaveLength(1);
    expect(spy.mock.calls[0]).toEqual(['a', synthEv]);
  });
});
