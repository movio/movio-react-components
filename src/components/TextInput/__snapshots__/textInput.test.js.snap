// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text-input-basic 1`] = `
<fieldset
  className="container"
>
  <input
    className="input"
    disabled={false}
    id="sample-input"
    onChange={[Function]}
    placeholder="..."
    type="text"
    value=""
  />
</fieldset>
`;

exports[`text-input-custom-classname 1`] = `
<fieldset
  className="container"
>
  <input
    className="test-class input"
    disabled={false}
    id="sample-input"
    onChange={[Function]}
    placeholder="..."
    type="text"
    value=""
  />
</fieldset>
`;

exports[`text-input-default-classname 1`] = `
<fieldset
  className="container"
>
  <input
    className="input"
    disabled={false}
    id="sample-input"
    onChange={[Function]}
    placeholder="..."
    type="text"
    value=""
  />
</fieldset>
`;

exports[`text-input-label 1`] = `
<fieldset
  className="container"
>
  <label
    className="label"
    htmlFor="sample-input"
  >
    Label
  </label>
  <input
    className="input"
    disabled={false}
    id="sample-input"
    onChange={[Function]}
    placeholder="..."
    type="text"
    value=""
  />
</fieldset>
`;
